text_generation/causal_lm/cpp/generate_pipeline/generate_pipeline.hpp

```diff
@@ -0,0 +1,453 @@
+        if (inputs_data[batch_offset + sequence_length - 1] != pad_token)
+            continue;
+
+        size_t pad_tokens_number = 0;
+        for (int i = sequence_length - 1; i >= 0; i--) {
+            const size_t token_offset = batch_offset + i;
+
+            if (inputs_data[token_offset] == pad_token)
+                continue;
+
+            if (pad_tokens_number == 0)
+                pad_tokens_number = sequence_length - i - 1;
+
+            std::swap(inputs_data[token_offset], inputs_data[token_offset + pad_tokens_number]);
+            std::swap(attention_mask_data[token_offset], attention_mask_data[token_offset + pad_tokens_number]);
+        }
+    }
+
+    return {input_ids, attention_mask};
+}
+
+void update_position_ids(ov::Tensor& position_ids, const ov::Tensor& attention_mask) {
+    const size_t batch_size = attention_mask.get_shape().at(0);
+    const size_t seq_length = attention_mask.get_shape().at(1);
+    position_ids.set_shape({batch_size, 1});
+
+    for (size_t batch = 0; batch < batch_size; batch++) {
+        int64_t* start = attention_mask.data<int64_t>() + batch * seq_length;
+        position_ids.data<int64_t>()[batch] = std::accumulate(start, start + seq_length, 0);
+    }
+}
+
+void initialize_position_ids(ov::Tensor& position_ids, const ov::Tensor& attention_mask) {
+    const size_t batch_size = attention_mask.get_shape()[0];
+    const size_t seq_length = attention_mask.get_shape()[1];
+
+    const int64_t* attention_mask_data = attention_mask.data<int64_t>();
+    int64_t* position_ids_data = position_ids.data<int64_t>();
+
+    for (size_t batch = 0; batch < batch_size; batch++) {
+        size_t sum = 0;
+        for (size_t i = 0; i < seq_length; i++) {
+            const size_t element_offset = batch * seq_length + i;
+            position_ids_data[element_offset] = sum;
+            if (attention_mask_data[element_offset] == 1) {
+                sum += 1;
+            }
+        }
+    }
+}
+
+ov::Tensor init_attention_mask(ov::Tensor& position_ids) {
+    auto shape = position_ids.get_shape();
+    auto attention_mask = ov::Tensor{position_ids.get_element_type(), shape};
+    std::fill_n(attention_mask.data<int64_t>(), shape[0] * shape[1], 1);
+    return attention_mask;
+}
+
+ov::Tensor extend_attention(ov::Tensor attention_mask) {
+    auto shape = attention_mask.get_shape();
+    auto batch_size = shape[0];
+    auto seq_len = shape[1];
+
+    ov::Tensor new_atten_mask = ov::Tensor{attention_mask.get_element_type(), {batch_size, seq_len + 1}};
+    auto old_data = attention_mask.data<int64_t>();
+    auto new_data = new_atten_mask.data<int64_t>();
+    for (size_t batch = 0; batch < batch_size; ++batch) {
+        std::memcpy(new_data + batch * (seq_len + 1), old_data + batch * seq_len, seq_len * sizeof(int64_t));
+        new_data[batch * (seq_len + 1) + seq_len] = 1;
+    }
+    return new_atten_mask;
+}
+
+class LLMPipeline {
+    ov::InferRequest m_model_runner;
```

Comment:
- how it is supposed to reset_state?
- can we add example when pipe is called multiple times?

Reply:
Currently we reset state always after generate/pipe(). But if user did `pipe.start_conversation` then state is saved between calls. Also user always can manually do `pipe.reset_state()`.

Added more examples in chat sample.